### ISSUE TITLE
Build fails under alpine with musl libc

### ISSUE THREAD

**Issue by eisensheng** [Snaipe/Criterion#143]
Building Criterion under alpine linux with musl libc fails

```
[ 37%] Building C object CMakeFiles/criterion.dir/src/core/abort.c.o
In file included from …/Criterion/src/compat/internal.h:56:0,
                 from …/Criterion/src/compat/process.h:28,
                 from …/Criterion/src/core/worker.h:29,
                 from …/Criterion/src/io/event.h:28,
                 from …/Criterion/src/core/abort.c:30:
/usr/include/sys/signal.h:1:2: error: #warning redirecting incorrect #include <sys/signal.h> to <signal.h> [-Werror=cpp]
 #warning redirecting incorrect #include <sys/signal.h> to <signal.h>
  ^
In file included from …/Criterion/src/compat/internal.h:57:0,
                 from …/Criterion/src/compat/process.h:28,
                 from …/Criterion/src/core/worker.h:29,
                 from …/Criterion/src/io/event.h:28,
                 from …/Criterion/src/core/abort.c:30:
/usr/include/sys/fcntl.h:1:2: error: #warning redirecting incorrect #include <sys/fcntl.h> to <fcntl.h> [-Werror=cpp]
 #warning redirecting incorrect #include <sys/fcntl.h> to <fcntl.h>
  ^
cc1: all warnings being treated as errors
make[2]: *** [CMakeFiles/criterion.dir/src/core/abort.c.o] Error 1
CMakeFiles/criterion.dir/build.make:62: recipe for target 'CMakeFiles/criterion.dir/src/core/abort.c.o' failed
make[1]: *** [CMakeFiles/criterion.dir/all] Error 2
CMakeFiles/Makefile2:106: recipe for target 'CMakeFiles/criterion.dir/all' failed
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

The offending file in musl libc:
https://git.musl-libc.org/cgit/musl/tree/include/sys/fcntl.h
```C
#warning redirecting incorrect #include <sys/fcntl.h> to <fcntl.h>
#include <fcntl.h>
```

It isn't clear why `sys/fcntl.h` is used here since even the man page for `fnctl` states that the correct header file would be `fcntl.h`.

**Comment by Snaipe** [comment]
Okay, I'll fix that once I get back. Is this the only issue that makes building with musl impossible?

**Comment by eisensheng** [comment]
> Okay, I'll fix that once I get back. Is this the only issue that makes building with musl impossible?

I gave it a try and fixed it locally to see what comes next. But it appears that ``fopencookie`` and friends are missing entirely from musl libc. So it seems to be quite the endeavor to compile Criterion with musl libc.

**Comment by Snaipe** [comment]
Le 22 août 2016 23:32, "Arthur Skowronek" <notifications@github.com> a
écrit :
>>
>> Okay, I'll fix that once I get back. Is this the only issue that makes
building with musl impossible?
>
> I gave it a try and fixed it locally to see what comes next. But it
appears that fopencookie and friends are missing entirely from musl libc.
So it seems to be quite the endeavor to compile Criterion with musl libc.
>

I'll take a look into properly building it with musl.

The part of code in question is a nonportable mess now that I take a look
at it again. Here funopen & fopencookie are both used to reimplement the
functionality of open_memstream, which musl has.

The code should test for features rather than the platform.

**Comment by Snaipe** [comment]
This has now been fixed with 805005a, 595b460, and 4f5b85d.

However, criterion hangs currently on alpine linux due to the nanomsg fork patch.

Incidentally, BoxFort also fails to handle some specifics of the musl library, so it happens to segfault on the features/boxfort-sandboxing branch.

Further development will be documented on Snaipe/Boxfort#1.

**Comment by qaqland** [comment]
any update on running with alpine linux, have you try it later?
 @eisensheng
